### PR TITLE
[JUJU-4475] Add autocert cache domain service

### PR DESCRIPTION
**Pull request by nvinuesa**
The new domain service (simply a proxy to the key value autocert cache) is now used on the httpserver worker.

Also, as a bonus, we remove the entire autocert legacy state.

Nit: remove unused Logger interface on external controllers domain service and move to debug the logging of acme's `GetCertificate()` error, which is not an error on juju's side.

## Checklist

*If an item is not applicable, use `~strikethrough~`.*

- [X] Code style: imports ordered, good names, simple structure, etc
- [X] Comments saying why design decisions were made
- [X] Go unit tests, with comments saying what you're testing
- [ ] ~[Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
- [ ] ~[doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

## QA steps

The main scenario is to bootstrap a controller using `autocert-dns-name` set to a domain we control, on which we'll have to add the controllers IP address to the dns entry. Then we just `openssl` client request the domain to verify that the cert is signed by let's encrypt and not juju-ca.

This was done thanks to @tlm 's help.

```
juju bootstrap aws c --config logging-config="juju.worker.servicefactory.autocertcache=TRACE" --config autocert-dns-name=ipv6galore.tlm.dev

openssl s_client -connect ipv6galore.tlm.dev:443
CONNECTED(00000003)
depth=2 C = US, O = Internet Security Research Group, CN = ISRG Root X1
verify return:1
depth=1 C = US, O = Let's Encrypt, CN = R3
...
```

### CODE REVIEW

File: domain/autocert/service/service.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+func NewService(st State, logger Logger) *Service {
+	return &Service{
+		st:     st,
+		logger: logger,
+	}
+}
+
+// Put implements autocert.Cache.Put.
+func (s *Service) Put(ctx context.Context, name string, data []byte) error {
+	s.logger.Warningf("storing autocert %s with contents '%+v' in the autocert cache", name, string(data))
```
Comment (tlm) [inline domain/autocert/service/service.go:44]: This should be trace I feel and we should drop the data logging.

File: domain/autocert/service/service.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	Delete(ctx context.Context, name string) error
+}
+
+// Logger facilitates emitting log messages.
+type Logger interface {
+	Tracef(string, ...interface{})
+	Warningf(string, ...interface{})
+}
+
+// Service provides the API for working with autocert cache.
```
Comment (tlm) [inline domain/autocert/service/service.go:28]: Nit, add comment saying that this service implements the cache interface as well?

File: domain/autocert/state/state.go [hunk 3]
```diff
@@ -85,7 +86,8 @@ WHERE  name = $M.name`
 		return errors.Trace(tx.Query(ctx, s, sqlair.M{"name": name}).Get(&row))
 	}); err != nil {
 		if err.Error() == sql.ErrNoRows.Error() {
-			return nil, errors.Annotatef(errors.NotFound, "autocert %s", name)
+			return nil, autocert.ErrCacheMiss
```
Comment (tlm) [inline domain/autocert/state/state.go:89]: I think this is the wrong place to place this.

My reasoning is that it's not the state layer that is implementing the autocert cache interface. It is the service that is implementing this. I think returning NotFound here is the correct response.

Then in the service layer we can do a `if errors.Is(err, errors.NotFound) {`

File: domain/autocert/state/state_test.go [hunk 4]
```diff
@@ -80,7 +80,7 @@ func (s *stateSuite) TestRetrieveNoCert(c *gc.C) {
 
 	// Retrieve an arbitrary non existent cert.
 	_, err := st.Get(context.Background(), "cert1")
-	c.Assert(err, gc.ErrorMatches, "autocert cert1: not found")
+	c.Assert(err, gc.ErrorMatches, "acme/autocert: certificate cache miss")
```
Comment (tlm) [inline domain/autocert/state/state_test.go:83]: I don't like asserting error strings as they are not meant for machine consumption. We should assert on type.

`c.Assert(err, jc.ErrorIs, errors.NotFound)`

File: domain/autocert/state/state_test.go [hunk 5]
```diff
@@ -277,5 +277,5 @@ Hn+GmxZA
 
 	// Retrieve the non-existent cert.
 	_, err = st.Get(context.Background(), "cert1")
-	c.Assert(err, gc.ErrorMatches, "autocert cert1: not found")
+	c.Assert(err, gc.ErrorMatches, "acme/autocert: certificate cache miss")
```
Comment (tlm) [inline domain/autocert/state/state_test.go:280]: Same comment as above around asserting type not string.

**Review by tlm — APPROVED** [review]
Just a few comments from me but LGTM otherwise.

File: domain/autocert/state/state_test.go (hunk 4 above)
Comment (nvinuesa) [reply domain/autocert/state/state_test.go]: absolutely, done!

File: domain/autocert/state/state.go (hunk 3 above)
Comment (nvinuesa) [reply domain/autocert/state/state.go]: yeah, I had think about this and then went on for state. Moved it to service layer

File: domain/autocert/service/service.go (hunk 2 above)
Comment (nvinuesa) [reply domain/autocert/service/service.go]: done

File: domain/autocert/service/service.go (hunk 1 above)
Comment (nvinuesa) [reply domain/autocert/service/service.go]: forgot to set it back to trace, done

File: domain/autocert/state/state_test.go (hunk 5 above)
Comment (nvinuesa) [reply domain/autocert/state/state_test.go]: done

File: domain/autocert/service/service.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+// Put implements autocert.Cache.Put.
+func (s *Service) Put(ctx context.Context, name string, data []byte) error {
+	s.logger.Tracef("storing autocert %s with contents '%+v' in the autocert cache", name, string(data))
+	return s.st.Put(ctx, name, data)
+}
+
+// Get implements autocert.Cache.Get.
+func (s *Service) Get(ctx context.Context, name string) ([]byte, error) {
```
Comment (manadart) [inline domain/autocert/service/service.go:52]: This style was mentioned in London. I am of the opinion that we should not conflate domain services with single entities, even though they might be in some cases.

So `entity.Get` makes sense, but `service.Get` is ambiguous. I would prefer to be explicit with naming. Will socialise this.

Comment (nvinuesa) [reply domain/autocert/service/service.go:52]: I 100% agree with you on the general case, but on this particular case the Service must satisfy the interface `autocert.Cache` (third party lib).

**Pull request merged**